main.cpp

```diff
@@ -38,8 +39,275 @@
+
+size_t LLVMFuzzerCustomMutator(uint8_t *fuzz_data, size_t size, size_t max_size,
+                               unsigned int seed)
+{
+    // A minimum size of 9 prevents hrp_maxlen <= 0 and data_maxlen <= 0.
+    if (size < 9)
+        return initial_input(fuzz_data, size, max_size);
+
+    // Interpret fuzz input as string
+    std::string input(reinterpret_cast<char*>(fuzz_data), size);
+
+    // Attempt to bech32 decode the input
+    bech32::DecodeResult decoded = bech32::Decode(input, bech32::CharLimit::CUSTOM_MUTATOR);
+    if (decoded.encoding != bech32::Encoding::BECH32) {
+        // Decoding failed, this should only happen when starting from
+        // an empty corpus.
+        return initial_input(fuzz_data, size, max_size);
+    }
+
+    auto data = decoded.data;
+    auto hrp = decoded.hrp;
+
+    size_t data_maxlen = input.size() > 8 ? std::min(input.size() - 8, static_cast<size_t>(65)) : 0;
+    size_t hrp_maxlen = input.size() > 6 ? std::min(input.size() - 6, static_cast<size_t>(83)) : 0;
+
+    // Mutate either the hrp or data
+    std::srand(seed);
+    switch (std::rand() % 2) {
+        case 0: { // Mutate hrp
+            if (hrp_maxlen > 0) {
+                // Make sure we have a buffer that's large enough for mutation
+                std::vector<uint8_t> hrp_buffer(hrp.begin(), hrp.end());
+                // Reserve enough space for the maximum mutation size
+                hrp_buffer.resize(hrp_maxlen);
+
+                // Mutate the buffer
+                size_t new_len = LLVMFuzzerMutate(hrp_buffer.data(),
+                                                 hrp.size(),
+                                                 hrp_maxlen - 1);
+
+                // Ensure minimum length and validate - at least 1 character
+                new_len = std::max(new_len, static_cast<size_t>(1));
+
+                // Convert back to string with the new length
+                hrp = std::string(reinterpret_cast<char*>(hrp_buffer.data()), new_len);
+
+                // Sanitize hrp - ensure only valid ASCII characters (33-126) and not uppercase
+                for (char& c : hrp) {
+                    if (c < 33 || c > 126) {
+                        c = 'a' + (c % 26); // Replace with a valid lowercase letter
+                    } else if (c >= 'A' && c <= 'Z') {
+                        c = c - 'A' + 'a'; // Convert uppercase to lowercase
+                    }
+                }
+            }
+            break;
+        }
+        case 1: { // Mutate data
+            if (data_maxlen > 0) {
+                // Make sure we have a buffer that's large enough for mutation
+                data.resize(std::max(data.size(), data_maxlen));
+
+                size_t datalen = data.size();
+                size_t new_len = LLVMFuzzerMutate(data.data(),
+                                                 std::min(datalen, data_maxlen),
+                                                 data_maxlen);
```

Comment:
I don't think this is doing the right thing...

1. `data` get's resized to `data_maxlen`.
2. `datalen` gets set to `data_maxlen`.
3. `LLVMFuzzerMutate` is called with `data_maxlen`.

Shouldn't `LLVMFuzzerMutate` be called with the original `data` size, not the resized size?